src/nvim/event/pty_process.c

```diff
@@ -20,92 +20,39 @@
 
 #include <uv.h>
 
-#include "nvim/func_attr.h"
-#include "nvim/os/job.h"
-#include "nvim/os/job_defs.h"
-#include "nvim/os/job_private.h"
-#include "nvim/os/pty_process.h"
-#include "nvim/memory.h"
+#include "nvim/lib/klist.h"
+
+#include "nvim/event/loop.h"
+#include "nvim/event/rstream.h"
+#include "nvim/event/wstream.h"
+#include "nvim/event/process.h"
+#include "nvim/event/pty_process.h"
+#include "nvim/log.h"
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
-# include "os/pty_process.c.generated.h"
+# include "event/pty_process.c.generated.h"
 #endif
 
-typedef struct {
-  struct winsize winsize;
-  uv_pipe_t proc_stdin, proc_stdout, proc_stderr;
-  int tty_fd;
-} PtyProcess;
-
-void pty_process_init(Job *job) FUNC_ATTR_NONNULL_ALL
-{
-  PtyProcess *ptyproc = xmalloc(sizeof(PtyProcess));
-  ptyproc->tty_fd = -1;
-
-  if (job->opts.writable) {
-    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stdin, 0);
-    ptyproc->proc_stdin.data = NULL;
-  }
-
-  if (job->opts.stdout_cb) {
-    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stdout, 0);
-    ptyproc->proc_stdout.data = NULL;
-  }
-
-  if (job->opts.stderr_cb) {
-    uv_pipe_init(uv_default_loop(), &ptyproc->proc_stderr, 0);
-    ptyproc->proc_stderr.data = NULL;
-  }
-
-  job->proc_stdin = (uv_stream_t *)&ptyproc->proc_stdin;
-  job->proc_stdout = (uv_stream_t *)&ptyproc->proc_stdout;
-  job->proc_stderr = (uv_stream_t *)&ptyproc->proc_stderr;
-  job->process = ptyproc;
-}
-
-void pty_process_destroy(Job *job) FUNC_ATTR_NONNULL_ALL
-{
-  xfree(job->opts.term_name);
-  xfree(job->process);
-  job->process = NULL;
-}
-
-static bool set_pipe_duplicating_descriptor(int fd, uv_pipe_t *pipe)
-  FUNC_ATTR_NONNULL_ALL
-{
-  int fd_dup = dup(fd);
-  if (fd_dup < 0) {
-    ELOG("Failed to dup descriptor %d: %s", fd, strerror(errno));
-    return false;
-  }
-  int uv_result = uv_pipe_open(pipe, fd_dup);
-  if (uv_result) {
-    ELOG("Failed to set pipe to descriptor %d: %s",
-         fd_dup, uv_strerror(uv_result));
-    close(fd_dup);
-    return false;
-  }
-  return true;
-}
-
 static const unsigned int KILL_RETRIES = 5;
 static const unsigned int KILL_TIMEOUT = 2;  // seconds
 
-bool pty_process_spawn(Job *job) FUNC_ATTR_NONNULL_ALL
+bool pty_process_spawn(PtyProcess *ptyproc)
+  FUNC_ATTR_NONNULL_ALL
 {
-  int master;
-  PtyProcess *ptyproc = job->process;
-  ptyproc->winsize = (struct winsize){job->opts.height, job->opts.width, 0, 0};
+  Process *proc = (Process *)ptyproc;
+  uv_signal_start(&proc->loop->children_watcher, chld_handler, SIGCHLD);
+  ptyproc->winsize = (struct winsize){ptyproc->height, ptyproc->width, 0, 0};
   struct termios termios;
   init_termios(&termios);
   uv_disable_stdio_inheritance();
-
+  int master;
   int pid = forkpty(&master, NULL, &termios, &ptyproc->winsize);
 
   if (pid < 0) {
+    ELOG("forkpty failed: %s", strerror(errno));
     return false;
   } else if (pid == 0) {
-    init_child(job);
+    init_child(ptyproc);
     abort();
   }
 
```

Comment:
Adding here a `goto error;` leads to a seg fault when using `:term`, because later on the uninitialized callback pointer `internal_close_cb` is used in `pty_process_close()`.